### ISSUE TITLE
Implement `PrimaryKey` on `AssetInfo` directly

### ISSUE THREAD

**Issue by CyberHoward** [mars-protocol/cw-asset#20]
I suggest to implement the cw-storage-plus `PrimaryKey` trait on `AssetInfo` directly instead of using an intermediary type `AssetInfoKey`. This allows the type to be easily used in maps and specifically multi-index maps.

**Comment by larry0x** [comment]
This is impossible due to Rust lifetimes.

In short, for a native coin such as `uosmo`, we want to have the bytearray `b"native:uosmo"` as key, while for a CW20 with address `osmo1...` we want `b"cw20:osmo1..."`.

To implement this we need to something like this:

```rust
impl<'a> PrimaryKey<'a> for AssetInfo {
    type Prefix = ();
    type SubPrefix = ();
    type Suffix = Self;
    type SuperSuffix = Self;

    fn key(&self) -> Vec<Key> {
        let s: String = self.to_string(); // generate the string representation of the asset info
        vec![Key::Ref(s.as_bytes())]      // convert the string to binary
    }
}
```

But this doesn't work! Because at the end of `key` function, the string `s` goes out of scope, so it's reference (the `as_bytes`) becomes the dangling reference. The compiler does not allow this.

This means we cannot construct the string inside the `key` function. The type to be implemented `PrimaryKey` must hold an _owned_ copy of the string itself.

If you have a better solution, please definitely suggest.

cc @grod220 who discovered this problem

**Comment by CyberHoward** [comment]
Cool, that's an interesting problem. 

I ran into this before and [hacked a way around it](https://docs.rs/abstract-os/0.6.2/src/abstract_os/objects/module.rs.html#55-75) by using `str`s

Applying the same here gives me something like: 
```rust
impl<'a> PrimaryKey<'a> for AssetInfo {
    type Prefix = ();
    type SubPrefix = ();
    type Suffix = Self;
    type SuperSuffix = Self;

    fn key(&self) -> Vec<Key> {
        let mut keys = vec![];
        match &self {
            AssetInfo::Cw20(addr) => {
                keys.extend("cw20:".key());
                keys.extend(addr.key());
            },
            AssetInfo::Native(denom) => {
                keys.extend("native:".key());
                keys.extend(denom.key());
            }
            AssetInfo::Cw1155(addr,id ) => {
                keys.extend("cw1155:".key());
                keys.extend(addr.key());
                keys.extend(":".key());
                keys.extend(id.key());
            }
        };
        keys
    }
}

```
Which compiles!

**Comment by larry0x** [comment]
Implemented in #21
